src/middlewares.ts

```diff
@@ -15,37 +15,85 @@ import {
 	touchCommit,
 } from './api';
 
-export async function determineCommitHash(req: any, res: any, next: any) {
-	const isHashInSession = !!req.session.commitHash;
-	const isHashSpecified = req.query && (req.query.hash || req.query.branch);
-	let commitHash;
+function assembleSubdomainUrlForHash(req: any, commitHash: string) {
+    const protocol = (req.secure ? 'https' : 'http');
 
-	if (isHashInSession && !isHashSpecified) {
-		next();
-		return;
-	}
+    return protocol + '://hash-' + commitHash + '.' + stripCommitHashSubdomainFromHost(req.headers.host);
+}
 
-	if (!isHashSpecified) {
-		res.send('Please specify a branch to load');
-		return;
-	}
+function stripCommitHashSubdomainFromHost(host: string) {
+    let segments = host.split('.'),
+        commitHashIndex = null;
+
+    for (let i = 0; i < segments.length; i += 1) {
+        if (0 === segments[i].indexOf('hash-')) {
+            commitHashIndex = i;
+            break;
+        }
+    }
+
+    if (null === commitHashIndex) {
+        return host;
+    }
+
+    return segments.slice(commitHashIndex + 1).join('.');
+}
+
+function getCommitHashFromSubdomain(req: any) {
+    const commitHash = _.find(
+        req.headers.host.split('.'),
+        function (hostSegment) {
+            return 0 === hostSegment.indexOf('hash-');
+        }
+    );
+
+    return commitHash.replace(/^hash-/, '');
```

Comment:
this could give us a type error when trying to call `replace` on `undefined` if there's no match.

we could take the regexp approach as well…

```js
function commitHashFromHost( host: string ): ?string {
	const match = /(?:^|.*?\.)hash-([a-f0-9]+)\./.exec( host );

	if ( ! match ) {
		return null;
	}

	const [ /* full match */, hash ] = match;
	return hash;
}
```

Reply:
is there a need to pass in `express.Request` here vs. a string, letting the caller unwrap it?

Reply:
I'm now using a shared regex pattern between the two utility functions.